folly/CMakeLists.txt

```diff
@@ -13,3 +13,31 @@ install(
 )
 
 add_subdirectory(experimental/exception_tracer)
+
+get_target_property(pic folly POSITION_INDEPENDENT_CODE)
```

Comment:
Is this always false by default?

I think we should strongly discourage building `libfolly` as a shared library in most circumstances.  When the cython build is enabled maybe we should compile the code twice: once as `libfolly.a` library to be used by downstream C++ dependencies, and a separate PIC build just for Cython users.  Perhaps it would be good to have the Cython extension directly include all of the folly code rather than linking against folly as a separate shared library.

That said, I'm not really that familiar with the current python code in folly and how it gets used to know if bundling all of folly directly into it makes sense in practice.  (This would potentially pose a challenge for dependencies that want to depend on folly both via the cython extension and directly as a C++ library.)

Reply:
I followed the example for singleton_thread_local_overload in top-level CMakeLists.txt.
I could look into, cherry-picking the objects required by the python extensions into code into the a new library; will see how far I get.

Reply:
I wouldn't bother with cherry-picking objects to create a new library, I was more just suggesting building two different versions of the existing library.

e.g., we could build `libfolly.a` and `libfolly_pic.a`, and the cython extension could link in `libfolly_pic.a`  You could probably do this relatively easily with something like

  add_library(folly_pic $<TARGET_OBJECTS:folly_base>)
  set_property(TARGET folly_pic PROPERTY POSITION_INDEPENDENT_CODE 1)

Reply:
I have updated the diff to include a folly_pic library; but it pretty much means compiling all the sources twice. (I tested the suggested code:
```
add_library(folly_pic $<TARGET_OBJECTS:folly_base>)
set_property(TARGET folly_pic PROPERTY POSITION_INDEPENDENT_CODE 1)
```
but this resulted in a "folly_pic" which was compiled without -fPIC; because this option was not applied to the creation of the objects in folly_base.

I have also read quite a bit further on the subject of position independent code; while it may have a performance impact on the compiled code, as I understanding Darwin enable PIC by default and its mandatory for iOS.

https://eli.thegreenplace.net/2011/11/03/position-independent-code-pic-in-shared-libraries/
(specifically "The costs of PIC")